piro/route.py

```diff
@@ -301,200 +252,40 @@ def sigma(self):
-        num_sites = sum(
-            self.target_entry.structure.composition.reduced_composition.as_dict().values()
-        )
-        _coeffs = []
-        for i in range(len(coeffs)):
-            p = self.reactions[rxn_label]["precursors"][i]
-            _coeffs.append(
-                coeffs[i]
-                / sum(p.structure.composition.reduced_composition.as_dict().values())
-                * num_sites
-            )
-        _coeffs = np.round(_coeffs, decimals=4)
-        report = " + ".join(
-            [
-                i + " " + j
-                for i, j in list(
-                    zip(
-                        [str(c) for c in _coeffs],
-                        [
-                            p.structure.composition.reduced_formula
-                            + "("
-                            + p.entry_id
-                            + ")"
-                            for p in self.reactions[rxn_label]["precursors"]
-                        ],
-                    )
-                )
-            ]
-        )
-        self.reactions[rxn_label]["summary"] = report
-        return report
-
-    def get_competing_phases(self, rxn_label, confine_to_icsd=True):
-
-        precursors = self.reactions[rxn_label]["precursors"]
-        precursor_ids = [i.entry_id for i in precursors]
-        _competing = []
-        _competing_rxe = []
+    def get_reactions(self):
 
-        for entry in self.entries:
-            if confine_to_icsd:
-                if not entry.data["icsd_ids"]:
-                    continue
-            if self.flexible_competition:
-                s1 = set(entry.composition.as_dict().keys())
-                s2 = set(self.target_entry.structure.composition.as_dict().keys())
-                if not (
-                    s1.issubset(s2)
-                    and (len(s2) - self.flexible_competition <= len(s1) <= len(s2))
-                ):
-                    continue
-            else:
-                if not set(self.target_entry.composition.as_dict().keys()).issubset(
-                    set(entry.structure.composition.as_dict().keys())
-                ):
-                    continue
-            if entry.entry_id in precursor_ids:
-                continue
-            if entry.entry_id == self.target_entry_id:
-                continue
-            competing_target_entry = entry
-
-            elts_precs = set()
-            for s in [
-                set(p.structure.composition.as_dict().keys()) for p in precursors
-            ]:
-                elts_precs = elts_precs.union(s)
-            if not set(entry.composition.as_dict().keys()).issubset(elts_precs):
-                # print(entry.composition)
+        for reaction_result in generate_reactions(
+                self.target_entry,
+                self.elts,
+                self.precursor_library,
+                self.allow_gas_release
+        ):
+            if reaction_result.label in self.reactions:
```

Comment:
Since `recommend_routes` calls `get_reactions` only when certain parameters are changed [here](https://github.com/TRI-AMDD/piro/blob/d2b593c56c35bd85d9faefcf531e4e47c23ff5bb/piro/route.py#L321), we should "wipe" `self.reactions` and run the updates below regardless. This would ensure, when, for example, temperature is changed, and `recommend_routes` is called again, the updated reaction energies are reflected in barriers.

Nevertheless, I'm realizing my original implementation of when/how get_reactions or updates are called is actually inefficient. So if we want to improve (in this or another PR):
- balanced reactions in principle need only be generated when `recommend_routes` is called for the first time, or when `allow_gas_release` is changed since the last call to `recommend_routes`
- updates of reaction energy, nucleation barrier and rxn summary should be computed on the first call or when temperature or pressure has changed since the last call. (and if we get into the weeds, we'd in fact need to do this only if a particular reaction has any gas input or output where temperature/pressure has the only effect for now in our implementation...)

I don't think we need to do this improvement in this PR, but let me know if you want to discuss in person.

Reply:
Another good catch, thanks for the thorough review. Yea I didn't do a lot of testing around consecutive use.

1) I'll make an update  by tomorrow to wipe self.reactions when we call get_reactions

2) For your further improvements. If it's ok, I'd prefer to move this improvement to another PR. You're talking about efficiency of consecutive calls to recommend_routes, whereas this PR focuses more on the efficiency of a single call.  And if we're talking about caching, this will only help local use. For the web app, we would require external cache like epitaxy and similarity. Adding an external cache for a lot of these reaction steps would require a larger PR. So maybe we should discuss in person.

Reply:
Yep, I completely agree. We should postpone the improvements to a future PR and just fix this one for now so that it functions as originally intended.

Reply:
1) updated the code to fix
2) created a new Jira issue. Regardless of how we go, I'll want to refactor the code more to accommodate